src/query/plan/operator.cpp

```diff
@@ -5222,4 +5223,133 @@ void IndexedJoin::IndexedJoinCursor::Reset() {
   pull_input_ = true;
 }
 
+std::vector<Symbol> HashJoin::ModifiedSymbols(const SymbolTable &table) const {
+  auto symbols = left_op_->ModifiedSymbols(table);
+  auto right = right_op_->ModifiedSymbols(table);
+  symbols.insert(symbols.end(), right.begin(), right.end());
+  return symbols;
+}
+
+bool HashJoin::Accept(HierarchicalLogicalOperatorVisitor &visitor) {
+  if (visitor.PreVisit(*this)) {
+    left_op_->Accept(visitor) && right_op_->Accept(visitor);
+  }
+  return visitor.PostVisit(*this);
+}
+
+WITHOUT_SINGLE_INPUT(HashJoin);
+
+namespace {
+
+class HashJoinCursor : public Cursor {
+ public:
+  HashJoinCursor(const HashJoin &self, utils::MemoryResource *mem)
+      : self_(self),
+        left_op_cursor_(self.left_op_->MakeCursor(mem)),
+        right_op_cursor_(self_.right_op_->MakeCursor(mem)),
+        hashtable_(mem),
+        right_op_frame_(mem) {
+    MG_ASSERT(left_op_cursor_ != nullptr, "HashJoinCursor: Missing left operator cursor.");
+    MG_ASSERT(right_op_cursor_ != nullptr, "HashJoinCursor: Missing right operator cursor.");
+  }
+
+  bool Pull(Frame &frame, ExecutionContext &context) override {
+    SCOPED_PROFILE_OP("HashJoin");
+
+    if (!hash_join_initialized_) {
+      // Pull all left_op_ frames
+      while (left_op_cursor_->Pull(frame, context)) {
+        ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.db_accessor,
+                                      storage::View::NEW);
+        auto left_value = self_.hash_join_condition_->expression2_->Accept(evaluator);
```

Comment:
why is left value with expression2 and not expression1?

Reply:
can you provide an example

Reply:
I’ve fixed this, but it shouldn’t cause issues anyway because HashJoin only handles `=` filters.

The `expression1_`/`expression2_` swap is because I followed Cartesian’s pretty print method:
bool PlanPrinter::PreVisit(query::plan::Cartesian &op) {
  WithPrintLn([&op](auto &out) {
    out << "* Cartesian {";
    utils::PrintIterable(out, op.left_symbols_, ", ", [](auto &out, const auto &sym) { out << sym.name(); });
    out << " : ";
    utils::PrintIterable(out, op.right_symbols_, ", ", [](auto &out, const auto &sym) { out << sym.name(); });
    out << "}";
  });
  Branch(*op.right_op_);
  op.left_op_->Accept(*this);
  return false;
}
As you can see, it prints the left symbols first, but then it prints right_op in the main branch.

Reply:
The left values have to be taken from `expression2_` and the right ones from `expression_1`, otherwise queries that use HashJoin return an empty result (e.g. the "Multiple match with WHERE x = y" tests in [cartesian.feature](https://github.com/memgraph/memgraph/blob/7b08af958086b123f07aa0c2cb51ccb270ceb0d4/tests/gql_behave/tests/memgraph_V1/features/cartesian.feature))